### PR TITLE
Add complex mandatory restrictions support

### PR DESCRIPTION
**Pull request by merkispavel**
# Issue

At the moment valhalla doesn't process `only_straight_on`, `only_left_turn`, `only_right_turn` complex restrictions(i.e with ways as via or conditions). This PR introduces the support of these restriction types.

## General idea
.. is to replace mandatory restriction of type `only_` with  prohibitory restrictions

pr

Python pseudocode would look like

```python
def replace_only_restriction(edges):
    for edge in outgoing_edges(edges[-2].endnode()):
        if edge != edges[-1]:
        add_prohibitory_restriction(edges[:-1] + [edge])

def handle_only_restriction(edges):
    for i in range(2, len(edges)):
        replace_only_restriction(edges[:i])
```

## Tricky parts of the solution
Let's say we have a graph
```
# only_straight_on restriction ABCD

A -> B -> C -> D
      \
       E
```
The algorithm will replace mandatory restriction `ABCD` with a prohibitory restriction `ABE`. However
- The edge `BE` might not be marked as `end_restriction`  but it must be
- The edge `EB`(opposite to the `BE`) might not be marked as `part_of_complex_restriction` but it must be https://github.com/valhalla/valhalla/pull/2796

So we set all these flags after the main function(i.e `build`) is called. The number of `only_` complex restrictions is relatively small so it shouldn't bring significant performance drop imho

## Tasklist

 - [x] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Generally use squash merge to rebase and clean comments before merging
 - [ ] Update the [changelog](CHANGELOG.md)

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: src/mjolnir/restrictionbuilder.cc [hunk 1]
```diff
@@ -239,17 +242,101 @@ std::deque<GraphId> GetGraphIds(GraphId& n_graphId,
...
+  return complex_restriction;
+};
+
+struct Result {
+  uint32_t forward_restrictions_count;
+  uint32_t reverse_restrictions_count;
+  std::vector<ComplexRestrictionBuilder> restrictions;
+  std::unordered_set<GraphId> part_of_restriction;
+};
+
```
Comment (merkispavel) [inline src/mjolnir/restrictionbuilder.cc:282]: Set needed flags for `end_restction`-edges

File: src/mjolnir/restrictionbuilder.cc [hunk 2]
```diff
@@ -239,17 +242,101 @@ std::deque<GraphId> GetGraphIds(GraphId& n_graphId,
...
+  GraphReader reader(config);
+  for (const auto& i : restrictions) {
+    GraphId tile_id = i.first;
+    auto tile = reader.GetGraphTile(tile_id);
+    if (!tile)
+      continue;
+
+    GraphTileBuilder tile_builder(reader.tile_dir(), tile_id, true);
+    for (auto restriction : i.second) {
+      tile_builder.AddForwardComplexRestriction(*restriction);
```
Comment (merkispavel) [inline src/mjolnir/restrictionbuilder.cc:305]: Add new prohibitory restriction that we couldn't handle within the main `build` function

File: src/mjolnir/graphtilebuilder.cc [hunk 3]
```diff
@@ -187,6 +187,45 @@ GraphTileBuilder::GraphTileBuilder(const std::string& tile_dir,
   lane_connectivity_builder_.reserve(n);
   std::copy(lane_connectivity_, lane_connectivity_ + n,
             std::back_inserter(lane_connectivity_builder_));
+
```
Comment (merkispavel) [inline src/mjolnir/graphtilebuilder.cc:190]: Needed to append forward restriction https://github.com/valhalla/valhalla/pull/2766/files#r570901414

**Comment by merkispavel** [comment]
PR is basically ready. But I found a bug in `GetGraphIds` function while testing the PR. Going to fix it asap to unblock the PR

File: src/mjolnir/restrictionbuilder.cc [hunk 4]
```diff
@@ -237,13 +357,11 @@ void build(const std::string& complex_restriction_from_file,
 
     size_t forward_count = 0, reverse_count = 0;
 
-    uint32_t id = tile_id.tileid();
     GraphId prevNode;
     for (uint32_t i = 0; i < tilebuilder.header()->nodecount(); i++) {
       NodeInfo& nodeinfo = tilebuilder.node_builder(i);
 
       // Go through directed edges and "enhance" directed edge attributes
-      const DirectedEdge* edges = tilebuilder.directededges(nodeinfo.edge_index());
```
Comment (merkispavel) [inline src/mjolnir/restrictionbuilder.cc:246]: unused

File: src/mjolnir/restrictionbuilder.cc [hunk 5]
```diff
@@ -237,13 +357,11 @@ void build(const std::string& complex_restriction_from_file,
 
     size_t forward_count = 0, reverse_count = 0;
 
-    uint32_t id = tile_id.tileid();
```
Comment (merkispavel) [inline src/mjolnir/restrictionbuilder.cc:240]: unused

File: src/mjolnir/restrictionbuilder.cc [hunk 6]
```diff
@@ -30,6 +31,43 @@ struct EdgeId {
   GraphId graph_id;
 };
 
+GraphId GetOpposingEdge(GraphReader& reader,
```
Comment (merkispavel) [inline src/mjolnir/restrictionbuilder.cc:34]: Copied it from the shortcut code https://github.com/valhalla/valhalla/blob/da75fe9b757123faf172511c25b54570f666d09f/src/mjolnir/shortcutbuilder.cc#L97

File: src/mjolnir/restrictionbuilder.cc [hunk 7]
```diff
@@ -271,143 +389,164 @@ void build(const std::string& complex_restriction_from_file,
...
-                    complex_restriction.set_end_month(td.end_month());
-                    complex_restriction.set_end_week(td.end_week());
-                  }
+                    for (size_t i = 0; i < end_node_tile->node(end_node)->edge_count(); ++i) {
+                      GraphId next_edge_id(end_node_tile->id().tileid(), end_node_tile->id().level(),
+                                           end_node_tile->node(end_node)->edge_index() + i);
+                      auto de = end_node_tile->directededge(next_edge_id);
+                      // reader, lock, from_node, edge_id, from_tile
+                      auto opp_id = GetOpposingEdge(reader, lock, end_node_tile, end_node, de);
+                      if (opp_id != last_edge_id && !de->is_shortcut() &&
```
Comment (merkispavel) [inline src/mjolnir/restrictionbuilder.cc:502]: wondering if we should skip platform & egress & transit like we do while matching the osmrestriction https://github.com/valhalla/valhalla/blob/da75fe9b757123faf172511c25b54570f666d09f/src/mjolnir/restrictionbuilder.cc#L67

**Comment by merkispavel** [comment]
Checked restrictions which are added by these changes to the Floridad tiles + rad tested: lgtm. Ready for review

File: src/mjolnir/restrictionbuilder.cc (hunk 7 above)
Comment (gknisely) [reply src/mjolnir/restrictionbuilder.cc:502]: @merkispavel yes we should skip them.  I would just run a quick numbers check(a before and after) on a small state to make sure all is good.

File: src/mjolnir/restrictionbuilder.cc [hunk 8]
```diff
@@ -271,143 +389,164 @@ void build(const std::string& complex_restriction_from_file,
...
+                      auto to_tile = reader.GetGraphTile(to_node);
+                      lock.unlock();
+                      auto to_node_info = to_tile->node(to_node);
+                      GraphId next_edge_id(to_tile->id().tileid(), to_tile->id().level(),
+                                           to_node_info->edge_index());
+                      for (size_t i = 0; i < to_node_info->edge_count(); ++i, ++next_edge_id) {
+                        auto de = to_tile->directededge(next_edge_id);
+                        // reader, lock, from_node, edge_id, from_tile
+                        auto opp_id = GetOpposingEdge(reader, lock, to_tile, to_node, de);
+                        if (opp_id != last_edge_id && !de->is_shortcut() &&
```
Comment (gknisely) [inline src/mjolnir/restrictionbuilder.cc:521]: @merkispavel don't forget to do it here too

**Review by gknisely — CHANGES_REQUESTED** [review]
@merkispavel commented on the PR.  Let me know if the numbers look good

File: src/mjolnir/restrictionbuilder.cc (hunk 7 above)
Comment (merkispavel) [reply src/mjolnir/restrictionbuilder.cc]: ty, done

**Comment by merkispavel** [comment]
@gknisely the numbers looks good 👍

**Review by gknisely — APPROVED** [review]

**Pull request merged**